fcrepo/fcrepo
# Return OCFL path in response

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3935

Alternative implementation to #2079 

Relies on: https://github.com/fcrepo/fcrepo-storage-ocfl/pull/58

# What does this Pull Request do?
The OCFL FileDetails object has a [getStorageRelativePath()](https://github.com/OCFL/ocfl-java/blob/main/ocfl-java-api/src/main/java/io/ocfl/api/model/FileDetails.java#L59-L66) function which returns the full path of the OCFL hidden Fedora header file when we read the headers. The related `fcrepo-storage-ocfl` PR passes this path into the storage layer ResourceHeaders, which are then transferred to the kernel level ResourceHeaders and up to the HTTP layer.

This does not show the storage specific prefix. At this level we only have an `objectSession` and (I think) we would need the entire repository (`io.ocfl.api.OcflRepository`) to get the storage (`io.ocfl.core.storage.OcflStorage`) to get the [`objectRootPath()`](https://github.com/OCFL/ocfl-java/blob/main/ocfl-java-core/src/main/java/io/ocfl/core/storage/OcflStorage.java#L182-L188).

~This method also does not have a way to stop this new information from being loaded. Because of the existing functionality, it requires a second touch of the file which (I imagine) will have some performance impacts.~
Edit: I have re-added the toggle property and have made changes in the `fcrepo-storage-ocfl` PR that hopefully reduces duplication and (by extension) performance impacts.

I have also stripped the path elements (the last 4 parts of the path) to make
`a3e/b5e/29e/a3eb5e29e71a3c6c8eec9ccb27622a454ec8b8f93393074d1e81223e18eca3f8/v1/content/.fcrepo/fcr-root.json`
to
`a3e/b5e/29e/a3eb5e29e71a3c6c8eec9ccb27622a454ec8b8f93393074d1e81223e18eca3f8`

# How should this be tested?

Build PRs, run Fedora, create an object, add the `fcrepo.ocfl.show_path=true` property and see the new `Fedora-Ocfl-Path` header.

For a container
```
> curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-container-1
HTTP/1.1 200 OK
Date: Tue, 10 Sep 2024 14:29:22 GMT
...
Last-Modified: Wed, 21 Aug 2024 17:11:11 GMT

Fedora-Ocfl-Path: 1ae/96b/c16/1ae96bc1695d8d30ebba20f010da0b9421afbd57cea05d2baa7b2d1f0717d35d

Link: <http://localhost:8080/rest/test-container-1>; rel="timegate"
...
Content-Type: text/turtle;charset=utf-8
Content-Length: 0
Server: Jetty(9.4.44.v20210927)
```

For a binary
```
> curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-container-1/a-binary
HTTP/1.1 200 OK
Date: Tue, 10 Sep 2024 14:33:50 GMT
...
Content-Type: image/jpeg
Accept-Ranges: bytes
Content-Disposition: attachment; size=2395723; creation-date="Tue, 10 Sep 2024 14:33:27 GMT"; modification-date="Tue, 10 Sep 2024 14:33:27 GMT"

Fedora-Ocfl-Path: a3e/b5e/29e/a3eb5e29e71a3c6c8eec9ccb27622a454ec8b8f93393074d1e81223e18eca3f8

Link: <http://localhost:8080/rest/test-container-1/a-binary/fcr:metadata>; rel="describedby"
...
Content-Length: 2395723
Server: Jetty(9.4.44.v20210927)
```

For a binary description
```
> curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-container-1/a-binary/fcr:metadata
HTTP/1.1 200 OK
Date: Tue, 10 Sep 2024 14:34:34 GMT
...
Last-Modified: Tue, 10 Sep 2024 14:33:27 GMT

Fedora-Ocfl-Path: a3e/b5e/29e/a3eb5e29e71a3c6c8eec9ccb27622a454ec8b8f93393074d1e81223e18eca3f8

Link: <http://localhost:8080/rest/test-container-1/a-binary>; rel="describes"
...
Content-Type: text/turtle;charset=utf-8
Content-Length: 0
Server: Jetty(9.4.44.v20210927)
```

For an Archival group
```
> curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/my-archival-group
HTTP/1.1 200 OK
Date: Tue, 10 Sep 2024 14:31:04 GMT
...
Last-Modified: Thu, 05 Sep 2024 16:29:17 GMT

Fedora-Ocfl-Path: 502/bb2/403/502bb2403dfb33624cf5b3cbf8e200fff81237fad89f0734b331ea35bd6a08c5

Link: <http://localhost:8080/rest/my-archival-group>; rel="timegate"
...
Content-Type: text/turtle;charset=utf-8
Content-Length: 0
Server: Jetty(9.4.44.v20210927)
```

and for an archival group member
```
> curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/my-archival-group/murphy-picture      
HTTP/1.1 200 OK
Date: Tue, 10 Sep 2024 14:35:51 GMT
...
Content-Type: image/jpeg
Accept-Ranges: bytes
Content-Disposition: attachment; size=2395723; creation-date="Thu, 5 Sep 2024 18:37:40 GMT"; modification-date="Thu, 5 Sep 2024 18:37:40 GMT"

Fedora-Ocfl-Path: 502/bb2/403/502bb2403dfb33624cf5b3cbf8e200fff81237fad89f0734b331ea35bd6a08c5

Link: <http://localhost:8080/rest/my-archival-group/murphy-picture/fcr:metadata>; rel="describedby"
...
Content-Length: 2395723
Server: Jetty(9.4.44.v20210927)
```

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
